node/core/pvf/common/src/worker/mod.rs

```diff
@@ -151,6 +179,87 @@ pub fn worker_event_loop<F, Fut>(
 	rt.shutdown_background();
 }
 
+/// Change root to be the artifact directory.
+#[cfg(target_os = "linux")]
+fn change_root(cache_path: &Path) -> Result<(), &'static str> {
+	use rand::{distributions::Alphanumeric, Rng};
+	use std::{ffi::CString, os::unix::ffi::OsStrExt, ptr};
+
+	const RANDOM_LEN: usize = 10;
+	let mut buf = Vec::with_capacity(RANDOM_LEN);
+	buf.extend(rand::thread_rng().sample_iter(&Alphanumeric).take(RANDOM_LEN));
+	let s = std::str::from_utf8(&buf)
+		.expect("the string is collected from a valid utf-8 sequence; qed");
+
+	let cache_path_str = match cache_path.to_str() {
+		Some(s) => s,
+		None => return Err("cache path is not valid UTF-8")
+	};
+	let cache_path_c = CString::new(cache_path.as_os_str().as_bytes()).unwrap();
+	let root_absolute_c = CString::new("/").unwrap();
+	// Append a random string to prevent races and to avoid dealing with the dir already existing.
+	let oldroot_relative_c = CString::new(format!("{}/oldroot-{}", cache_path_str, s)).unwrap();
+	let oldroot_absolute_c = CString::new(format!("/oldroot-{}", s)).unwrap();
+
+	// SAFETY: TODO
+	unsafe {
+		// 1. `unshare` the user and the mount namespaces.
+		if libc::unshare(libc::CLONE_NEWUSER) < 0 {
+			return Err("unshare user namespace")
+		}
+		if libc::unshare(libc::CLONE_NEWNS) < 0 {
+			return Err("unshare mount namespace")
+		}
+
+		// 2. `pivot_root` to the artifact directory.
+		gum::info!(target: LOG_TARGET, "1. {:?}", std::env::current_dir());
+		gum::info!(target: LOG_TARGET, "1.5. {:?}", std::fs::read_dir(".").unwrap().map(|entry| entry.unwrap().path()).collect::<Vec<PathBuf>>());
+		// TODO: Ensure that 'new_root' and its parent mount don't have shared propagation.
+		if libc::mount(ptr::null(), root_absolute_c.as_ptr(), ptr::null(), libc::MS_REC | libc::MS_PRIVATE, ptr::null()) < 0 {
+            return Err("mount MS_PRIVATE")
+		}
+		if libc::mount(
+			cache_path_c.as_ptr(),
+			cache_path_c.as_ptr(),
+			ptr::null(), // ignored when MS_BIND is used
+			libc::MS_BIND | libc::MS_REC | libc::MS_NOEXEC,
+			ptr::null(), // ignored when MS_BIND is used
+		) < 0
+		{
+			return Err("mount MS_BIND")
+		}
+		if libc::mkdir(oldroot_relative_c.as_ptr(), 0755) < 0 {
+			return Err("mkdir oldroot")
+		}
+		if libc::syscall(
+			libc::SYS_pivot_root,
+			cache_path_c.as_ptr(),
+			oldroot_relative_c.as_ptr(),
+		) < 0
+		{
+			return Err("pivot_root")
+		}
```

Comment:
If you call the pivot root syscall with both paths being `"."` then this `mkdir` should be unnecessary.